### PR TITLE
Support MacOS

### PR DESCRIPTION
**Pull request by WillChilds-Klein**
*Issue #, if available:* #41

# Notes
This commit adds functional and CI support for MacOS. The main blocker
on supporting MacOS was a runtime linking error caused by libaccp trying
to load a `libcrypto` it found in some `/usr` system directory on MacOS,
instead of the temporary directory that ACCP creates and restricts
`rpath` to during linking. The reason for this is that the `rpath` value
of `$ORIGIN` is specific to linux's linker, and is meaningless on MacOS.
So, the MacOS linker consulted a non-existent `rpath`, didn't find
anything, and fell back to system library directories. The fix is to
detect OS type at build time and use the appropriate value for `rpath`
(on MacOS that's `@loader_path`, see the "Locating External Resources"
section of [this documentation page][1]).  We also explicity set
`CMAKE_MACOSX_RPATH` on MacOS, which is documented [here][2].

To prevent regressions expanding RPATH beyond one entry, we add a
test check that asserts that libaccp only has a single RPATH entry
that's set to the appropriate value on all supported platforms.

The other pieces of this commit are related to CI. In order for the
coverate steps to pass, we need to install `lcov` and `gcovr`
dependencies on the GitHub Actions MacOS CI images. Lastly, now that
this runs on Mac the Apple JCE providers in our pre-baked properties
files are actually used (they're ignored on Linux, where they're
unavailable). This resulted in the same provider being installed twice,
causing SecurityPropertiesTester failures. We resolve this issue by
removing the class-specified entry in favor of the name-specified entry.
Because the [AppleProvider][3] is [not exported][4] from its module in
JDK 11, we cannot use the fully qualified class specification (see
comment in `etc/amazon-corretto-crypto-provider.security`) on JDK 9+. By
using only the name specification, there may be some issues on JDK8 on
Mac when using this providers file, but the impact should be minor -- no
regressions (because we didn't support MacOS before). Users are
encouraged to upgrade to more moder JDKs or manually modify/override the
supplied properties file.

This commit resolves [Issue #41][5].

[1]: https://developer.apple.com/library/archive/documentation/DeveloperTools/Conceptual/DynamicLibraries/100-Articles/DynamicLibraryDesignGuidelines.html#//apple_ref/doc/uid/TP40002013-SW21
[2]: https://cmake.org/cmake/help/latest/prop_tgt/MACOSX_RPATH.html
[3]: https://docs.oracle.com/javase/8/docs/technotes/guides/security/crypto/CryptoSpec.html#Provider
[4]: https://github.com/corretto/corretto-11/blob/develop/src/java.base/macosx/classes/module-info.java.extra
[5]: https://github.com/corretto/amazon-corretto-crypto-provider/issues/41

# Testing

## Manual Inspection of libaccp RPATH

On MacOS:

```
$ sw_vers
ProductName:    macOS
ProductVersion: 12.5.1
BuildVersion:   21G83

$ uname -a
Darwin 08f8bc77642e.ant.amazon.com 21.6.0 Darwin Kernel Version 21.6.0: Wed Aug 10 14:25:27 PDT 2022; root:xnu-8020.141.5~2/RELEASE_X86_64 x86_64

$ otool -l ./build/cmake/libamazonCorrettoCryptoProvider.dylib | grep -A2 ' cmd LC_RPATH'
          cmd LC_RPATH
      cmdsize 32
         path @loader_path (offset 12)
```

On Ubuntu:

```
root@817bfef6156a:/accp# cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.4 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.4 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal

root@817bfef6156a:/accp# chrpath -l ./build/cmake/libamazonCorrettoCryptoProvider.so
./build/cmake/libamazonCorrettoCryptoProvider.so: RUNPATH=$ORIGIN
```

On AL2:

```
$ cat /etc/os-release
NAME="Amazon Linux"
VERSION="2"
ID="amzn"
ID_LIKE="centos rhel fedora"
VERSION_ID="2"
PRETTY_NAME="Amazon Linux 2"
ANSI_COLOR="0;33"
CPE_NAME="cpe:2.3:o:amazon:amazon_linux:2:-:internal"
HOME_URL="https://amazonlinux.com/"
VARIANT="internal"

$ chrpath -l ./build/cmake/libamazonCorrettoCryptoProvider.so
./build/cmake/libamazonCorrettoCryptoProvider.so: RPATH=$ORIGIN
```

## RPATH Regression Testing Mechanism

Modified local code to include an extra RPATH entry.

Local changes:

```
$ git diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 0252625..da83da3 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -86,8 +86,6 @@ string(REPLACE ":" ";" TEST_CLASSPATH_LIST "${TEST_CLASSPATH}")

 # Needed as we abuse some of the test compile macros to test shared lib links
 set(CMAKE_POSITION_INDEPENDENT_CODE ON)
-# CMake's default RPath handling includes the full build directory path, which isn't needed. Turn it off.
-set(CMAKE_SKIP_RPATH TRUE)
 set(GENERATE_HASHERS ${CMAKE_CURRENT_SOURCE_DIR}/build-tools/bin/generate-java-hash-spi)
 set(GENERATED_JAVA_DIR ${CMAKE_CURRENT_BINARY_DIR}/generated-java/com/amazon/corretto/crypto/provider)
 set(JNI_HEADER_DIR ${CMAKE_CURRENT_BINARY_DIR}/generated-include)
```

On MacOS:

```
$ ./gradlew clean && ./gradlew test
...
/Library/Developer/CommandLineTools/usr/bin/make  -f CMakeFiles/check-libaccp-rpath.dir/build.make CMakeFiles/check-libaccp-rpath.dir/build
otool -l /Users/childw/workplace/github/corretto/amazon-corretto-crypto-provider/build/cmake/libamazonCorrettoCryptoProvider.dylib | grep 'cmd LC_RPATH' | wc -l | tr -d '[:blank:]' | xargs test 1 -eq
...
BUILD FAILED in 54s
3 actionable tasks: 2 executed, 1 up-to-date
```

On Ubuntu:

```
$ ./gradlew clean && ./gradlew test
...
make[3]: Entering directory '/accp/build/cmake'
objdump -x /accp/build/cmake/libamazonCorrettoCryptoProvider.so | grep -E 'RPATH|RUNPATH' | grep -v ':'
make[3]: *** [CMakeFiles/check-libaccp-rpath.dir/build.make:60: CMakeFiles/check-libaccp-rpath] Error 1
...
BUILD FAILED in 1m 28s
3 actionable tasks: 2 executed, 1 up-to-date
```

On AL2:

```
$ ./gradlew clean && ./gradlew test
...
make[3]: Entering directory `/local/home/childw/workplace/github/corretto/amazon-corretto-crypto-provider/build/cmake'
objdump -x /local/home/childw/workplace/github/corretto/amazon-corretto-crypto-provider/build/cmake/libamazonCorrettoCryptoProvider.so | grep -E 'RPATH|RUNPATH' | grep -v ':'
...
BUILD FAILED in 1m 31s
3 actionable tasks: 3 executed
```

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -86,8 +86,6 @@ string(REPLACE ":" ";" TEST_CLASSPATH_LIST "${TEST_CLASSPATH}")
 
 # Needed as we abuse some of the test compile macros to test shared lib links
 set(CMAKE_POSITION_INDEPENDENT_CODE ON)
-# CMake's default RPath handling includes the full build directory path, which isn't needed. Turn it off.
-set(CMAKE_SKIP_RPATH TRUE)
```
Comment (alexw91) [inline CMakeLists.txt:90]: If I remember correctly, before I added this line, the RPath of `libamazonCorrettoCryptoProvider.so` on my Ubuntu desktop contained both of the following when running `chrpath libamazonCorrettoCryptoProvider.so`:
1. `/home/aweibel/workspace/github/amazon-corretto-crypto-provider/build/awslc/bin/lib/libcrypto.so` (The path I where I was building ACCP on my desktop.)
2. `$ORIGIN`

Adding these 2 lines caused the 1st RPath value of the temporary build directory to be removed from the resulting  output, since only the 2nd RPath value is needed. Deleting these two lines **might** re-add the temporary build path's back into the libACCP on Linux platforms, and this should be checked before merging. Theoretically, if there's an extra RPath value containing a hard-coded directory from where our CI is configured to build and publish ACCP from, then any user on a host that is using ACCP could recreate that hard-coded path and put a different `libcrypto.so` file there that would loaded into any Java processes using ACCP on that host.

We may want to add a check during the build that ensures that only 1 RPath value is present in `libamazonCorrettoCryptoProvider.so`, and that it's value is `$ORIGIN`.

Comment (WillChilds-Klein) [reply CMakeLists.txt:90]: nice catch. looks like deleting this line does indeed introduce multiple rpath entries (albeit in the order we want)

```
root@817bfef6156a:/accp# cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.4 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.4 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
root@817bfef6156a:/accp# chrpath -l ./build/cmake/libamazonCorrettoCryptoProvider.so
./build/cmake/libamazonCorrettoCryptoProvider.so: RUNPATH=$ORIGIN:/accp/build/awslc/bin/lib
```

agreed that rpath should have one entry and point to `$ORIGIN`. i'll walk this back and start looking at ways to prune the rpath programmatically on MacOS and also look into a build-time check that asserts our rpath requirements.

**Comment by WillChilds-Klein** [comment]
checks are failing here due to CI weirdness, and we need to address the issues that alex called out before merging. but we're not far from full functionality. I have a PR [here](https://github.com/WillChilds-Klein/amazon-corretto-crypto-provider/pull/4) with all checks passing on MacOS.

File: CMakeLists.txt (hunk 1 above)
Comment (WillChilds-Klein) [reply CMakeLists.txt]: ok, digging into it further, it looks like the root cause for lack of MacOS support before this commit was that rpath value of `$ORIGIN` doesn't mean anything on MacOS. As shown above, my prior approach included the aws-lc build dir in rpath, which i think is where `libcrypto.so` was being found in the initial draft of this PR. the fix was to use the MacOS equivalent of `$ORIGIN`, which is [`@loader_path`](https://developer.apple.com/library/archive/documentation/DeveloperTools/Conceptual/DynamicLibraries/100-Articles/DynamicLibraryDesignGuidelines.html#//apple_ref/doc/uid/TP40002013-SW21). I've confirmed that in this PR's current version, there's only one RPATH entry (see testing section in the PR overview) and its set appropriately for mac and linux.

i also looked into adding a build-time check for this fact, but my CMake-fu is rather weak when it comes to invoking compound commands. in bash with `-e` and `-pipefail` set, i think the checks would look something like this:

mac:

```
# check that there's only one rpath value (probably not the best way to say "expect only one matching line")
[[ $(echo 1) == $(otool -l ./path/to/libamazonCorrettoCryptoProvider.dylib | ' cmd LC_RPATH' | wc -l)

# check that it's set to @loader_path
otool -l ./path/to/libamazonCorrettoCryptoProvider.dylib | grep -A2 ' cmd LC_RPATH' | tail -1 | grep 'libamazonCorrettoCryptoProvider'
```

linux:

```
# check that there's only one rpath value and that it's set to @loader_path
chrpath -l ./path/to/libamazonCorrettoCryptoProvider.so | grep -P 'RUNPATH=\$ORIGIN$'
```

File: etc/amazon-corretto-crypto-provider.security [hunk 2]
```diff
@@ -46,7 +46,6 @@ security.provider.17=JdkSASL
 security.provider.18=sun.security.mscapi.SunMSCAPI
 security.provider.19=SunMSCAPI
 security.provider.20=apple.security.AppleProvider
-security.provider.21=Apple
```
Comment (alexw91) [inline etc/amazon-corretto-crypto-provider.security:49]: Was this change intended?

Comment (WillChilds-Klein) [reply etc/amazon-corretto-crypto-provider.security:49]: yes, there's an explanation in the last paragraph of the Notes/commit message.

File: CMakeLists.txt (hunk 1 above)
Comment (WillChilds-Klein) [reply CMakeLists.txt]: test check added in most recent revision.

File: etc/amazon-corretto-crypto-provider.security (hunk 2 above)
Comment (WillChilds-Klein) [reply etc/amazon-corretto-crypto-provider.security:49]: updated explanation:

> Now that this runs on Mac the Apple JCE providers in our pre-baked properties
> files are actually used (they're ignored on Linux, where they're unavailable).
> This resulted in the same provider being installed twice, causing
> SecurityPropertiesTester failures. We resolve this issue by removing the
> class-specified entry in favor of the name-specified entry.  Because the
> [AppleProvider][3] is [not exported][4] from its module in JDK 11, we cannot
> use the fully qualified class specification (see comment in
> `etc/amazon-corretto-crypto-provider.security`) on JDK 9+. By using only the
> name specification, there may be some issues on JDK8 on Mac when using this
> providers file, but the impact should be minor -- no regressions (because we
> didn't support MacOS before). Users are encouraged to upgrade to more moder
> JDKs or manually modify/override the supplied properties file.

[3]: https://github.com/corretto/corretto-11/blob/develop/src/java.base/macosx/classes/apple/security/AppleProvider.java
[4]: https://github.com/corretto/corretto-11/blob/develop/src/java.base/macosx/classes/module-info.java.extra

**Comment by SalusaSecondus** [comment]
Please update the main readme to list the new support for MacOS. Thanks.

File: README.md [hunk 3]
```diff
@@ -83,7 +83,7 @@ KeyFactory algorithms:
 # Compatibility & Requirements
 ACCP has the following requirements:
 * JDK8 or newer (This includes both OracleJDK and [Amazon Corretto](https://aws.amazon.com/corretto/))
-* 64-bit Linux running on x86_64 (also known as x64 or AMD64)
+* 64-bit Linux or MacOs running on x86_64 (also known as x64 or AMD64)
```
Comment (SalusaSecondus) [inline README.md:86]: It works on Linux and MacOS running on aarch64 also, doesn't it? :-)

Comment (WillChilds-Klein) [reply README.md:86]: presumably. we know that it works on aarch64 linux, but we haven't yet tested it yet on aarch64 MacOS as far as I know (I use an x86 mac, and it's unclear what the underlying architecture is for the MacOS jobs).

in any case, I wanted to keep the README guidance on architecture and OS hewed to whatever is available in maven until we publish an external release of ACCP 2.0. until then, maven-vended ACCP will still only be supported on x86 linux.

**Review by alexw91 — APPROVED** [review]
